lessons/misc/intro-git/lesson.md

```diff
@@ -0,0 +1,394 @@
+word `git` -- so expect to be typing those three letters a whole lot!
+
+```bash
+mkdir aboutme/
+cd aboutme/
+git init # creates the repository
+```
+
+If done correctly, Git will tell us that our repo has been initialized:
+
+```
+Initialized empty Git repository in /Users/Ahmed/Desktop/aboutme/.git/
+```
+
+## Editing and tracking files
+
+Now, let's create a file and get Git to track it:
+
+```bash
+touch bio.txt # creates empty file called bio.txt
+ls # check to see that file has been made
+```
+
+Let's check on the state of our repo. 
+
+```bash
+git status
+```
+
+`git status` is an incredibly useful command that returns a host of important
+information about our repo. It's worth running it very liberally whenever one
+is working within a Git repo just to make sure everything looks the way it's
+expected to. 
+
+```
+$ git status
+On branch master
+
+No commits yet
+
+Untracked files:
+  (use "git add <file>..." to include in what will be committed)
+
+        bio.txt
+
+nothing added to commit but untracked files present (use "git add" to track)
+```
+
+Right now, all we have to be concerned about is the 'Untracked files' section.
+Notice that `bio.txt` is listed down there -- it may even be color coded as
+red. This is because Git does not track everything in a folder willy-nilly, and
+will instead wait to be told that `bio.txt` is something the user would like to
+track.
+
+We can tell Git exactly that by using `git add`:
+
+```bash
+git add bio.txt
+git status
+```
+
+`git status` now lists `bio.txt` under 'Changes to be committed'. 
+A _commit_ is the Git term for a save point. Let's do just that
+using the `git commit` command:
+
+```bash
+git commit -m 'initial commit'
+git status
+```
+
+Here, we use the `-m` flag to denote the message we want to save alongside the
+edits we've made. Git will _insist_ that a message is added on -- commits
+without messages are not permitted!  This might seem a bit pushy, but it's
+really Git trying to save you from yourself in making sure each change is
+documented.
+
+Let's now edit `bio.txt` a bit and see how Git responds.  Open `bio.txt` with
+vim in the command line (or whatever your editor of choice may be):
+
+```bash
+vim bio.txt
+```
+
+Press 'i' to enter insert mode in vim, and then add:
+
+- Your name
+- Your program and year
+- Your favourite programming language
+- One programming language you'd like to learn one day
+
+Remember to then hit ESC and type ':wq' to exit vim. 
+
+Let's check the status of our repo:
+
+```bash
+git status
+```
+
+Git's noticed we've changed the file. Not only that, but it can tell us what's
+been changed since the last commit via the `git diff` command:
+
+```bash
+git diff bio.txt
+```
+
+If we're happy with our changes, we have to once again use `git add` to prime
+(or 'stage') this file before committing.
+
+```bash
+git add bio.txt
+git commit -m 'added bio'
+```
+
+Let's check on all the changes we've made so far!
+
+```bash
+git log
```

Comment:
As a repo becomes larger, I find the default output of `git log` to be a bit cumbersome, especially when looking for a specific change. There a a few flags that may be worth mentioning to 'pretty' it up. The full command I use is `git log --oneline --all --decorate --graph`, which I've bound to an alias called `glog`. Maybe a good way to reming folks about aliases and briefly mention that like many command line tools, git commands have a bunch of optional flags to customize their function